src/types/datetime.rs

```diff
@@ -23,21 +23,29 @@ use crate::ffi_ptr_ext::FfiPtrExt;
 use crate::instance::PyNativeType;
 use crate::types::any::PyAnyMethods;
 use crate::types::PyTuple;
-use crate::{Bound, IntoPy, Py, PyAny, Python};
+use crate::{Bound, IntoPy, Py, PyAny, PyErr, Python};
 use std::os::raw::c_int;
 #[cfg(feature = "chrono")]
 use std::ptr;
 
-fn ensure_datetime_api(_py: Python<'_>) -> &'static PyDateTime_CAPI {
+fn ensure_datetime_api(py: Python<'_>) -> PyResult<&'static PyDateTime_CAPI> {
     unsafe {
         if pyo3_ffi::PyDateTimeAPI().is_null() {
```

Comment:
Since this is called repeatedly, maybe we can try to minimize the number of times we call into `PyDateTimeAPI`, e.g.

```rust
let mut api = pyo3_ffi::PyDateTimeAPI();

if api.is_null() {
 PyDateTime_IMPORT();

 api = pyo3_ffi::PyDateTimeAPI();
 if api.is_null() {
  return Err(PyErr::fetch(py));
 }
}

Ok(&*api)
```

Reply:
Good idea. I decided to also use `<*const T>::as_ref`, I ended up with this:

```rust
    if let Some(api) = unsafe { pyo3_ffi::PyDateTimeAPI().as_ref() } {
        Ok(api)
    } else {
        unsafe {
            PyDateTime_IMPORT();
            pyo3_ffi::PyDateTimeAPI().as_ref()
        }
        .ok_or_else(|| PyErr::fetch(py))
    }
```

I'm quite happy with that!